shop.cpp

```diff
@@ -3,24 +3,57 @@
 
 Shop::Shop(int mMaxChairs)
     :mMaxChairs(mMaxChairs)
-    ,mTheBarber(this)
-    ,mBarberSemaphore(0)
-    ,mCustomersSemaphore(0)
 {
+    mBarberSemaphore = new Semaphore(1);
+    mCustomersSemaphore = new Semaphore(0);
+    mTheBarber = new Barber(this, mBarberSemaphore, mCustomersSemaphore);
+}
+
+Shop::~Shop()
+{
+    if (mTheBarber != nullptr)
+    {
+        delete mTheBarber;
+    }
+
+    if (mBarberSemaphore != nullptr)
+    {
+        delete mBarberSemaphore;
+    }
+
+    if (mCustomersSemaphore != nullptr)
+    {
+        delete mCustomersSemaphore;
+    }
+}
+
+void Shop::newCustomerArrives(std::string customerName)
+{
+    mShopMutex.lock();
+    mTheCustomers.emplace_back(new Customer(customerName, this, mBarberSemaphore, mCustomersSemaphore));
```

Comment:
How are the customer objects deleted?

Reply:
Who owns the customers? Is it reasonable for the shop to create the customers?

Reply:
You are right, it is more reasonable that the Customers are externally created (in main.cpp) and then added to the Shop. In this way, main should also handle Customers deletion.